### ISSUE TITLE
export <> noexport feels like the wrong distinction

### ISSUE THREAD

**Issue by annevk** [tabatkins/bikeshed#797]
Basically, it seems a spec has a "public interface" and a "private interface". However, often other specifications will still want to refer to the "private interface" for various reasons. Therefore noexport doesn't really help, since really everything should be exported and easily linkable.

Perhaps for the private stuff there should be some keyword and then if bikeshed detects usage in a normative context, it can flag it?

**Comment by annevk** [comment]
Paging @mikewest @jyasskin @domenic @zcorpan for potentially different viewpoints.

**Comment by annevk** [comment]
The only case for noexport is really HTML, where it explicitly takes terms from other specifications and puts them inline. In that case it's actually annoying that you can overwrite noexport with `<a spec=html>...</a>` as those terms really never ought to be found other than by HTML's own tooling.

**Comment by tabatkins** [comment]
Public vs private is on fact precisely what the export/noexport distinction captures, and you can indeed refer to any spec's unexported terms of you really want to. (The trigger is that you need to specify the `spec` attribute, to indicate that you're sure about exactly what spec your trying to pull a definition from.)

And no, HTML is definitely not the only usecase.

**Comment by annevk** [comment]
So maybe HTML needs something like "don't index at all" for terms that are not actually defining anything whatsoever?

Note that when folks add export/noexport to specs they don't really take into account whether the entry point is public or private. They just put export on stuff they want to use. So they're not used as intended...

**Comment by tabatkins** [comment]
Yeah, that's reasonable. Maybe an empty linking text (`data-lt=""`), like you use to indicate an `<a>` isn't an autolink? Trivial to add.

**Comment by zcorpan** [comment]
Or maybe we should change those `dfn`s in HTML to `a` with some new attribute so it's never picked up by bikeshed but gives the intended behavior in wattsi? (Needs some changes to wattsi and HTML though.)

**Comment by domenic** [comment]
My perspective, FWIW:

- I agree that export/noexport ~ public/private
- I agree that people aren't using it this way. Due to the fact that there are defaults (especially the noexport default for dfn), they're just slapping export on things as other specs ask for them, instead of thinking about their public API from first principles. If I were dictator of Bikeshed and had a time-travel machine, I'd go back and make one of the two mandatory on all dfns.
  - We could try to address this by adding a per-organization flag that warns on dfns that don't explicitly specify, and turn it on for WHATWG. But it would generate a lot of busywork for us, and wouldn't solve HTML.
- In general I would weakly prefer there to be no escape hatch to reach into another spec's private API. But I'm sure that would mess a lot of things up.
- It sounds like some of this is just working around https://github.com/tabatkins/bikeshed/issues/792 and https://github.com/tabatkins/bikeshed/issues/786 ? That is, if those were straightened out, then we wouldn't need to mark up our stuff as never-link-to-this, since most of the problem is around Bikeshed/Shepherd's heuristics picking these up for unknown reasons, not around people adding `spec="html"`?
- Idle thought: I wonder why https://streams.spec.whatwg.org/#index-defined-elsewhere doesn't use `<dfn>` but HTML's https://html.spec.whatwg.org/multipage/infrastructure.html#dependencies does. They seem semantically the same.

**Comment by jyasskin** [comment]
I agree with @annevk in seeing at least 3 categories of definitions:

1. public: designed to be called from other specs.
2. descriptive: only designed to be "called" from the local spec, but can be used as part of another spec's explanatory text.
3. private: only designed to be used from the local spec. For example, some locally-used terms might be too short to be used from elsewhere.

**Comment by jyasskin** [comment]
I think it's essential to have a way to locally override a remote spec's decisions, in order to make progress while we get that remote spec fixed. I'd be very happy if every one of those overrides caused a warning in the local spec that it's not stable until all such overrides are fixed.

**Comment by tabatkins** [comment]
Re: people not using it this way: Not much I can do about that. I agree that the export/noexport semantics aren't expressed well, and I'd redo it if possible. Given the way the defaults work, maybe I *can* just redo it, and use words that are more evocative of purpose?

> We could try to address this by adding a per-organization flag that warns on dfns that don't explicitly specify, and turn it on for WHATWG. But it would generate a lot of busywork for us, and wouldn't solve HTML.

If your first reaction is that it'll create a lot of "busywork" even tho you're the one suggesting it, then it's probably not a good idea. ^_^

> In general I would weakly prefer there to be no escape hatch to reach into another spec's private API. But I'm sure that would mess a lot of things up.

Definitely not.  For one, as jyasskin says, you need a way to route around mistakes. For two, I can't stop people from just manually linking anyway, and that just results in greater fragility (no error-checking if you misspell something, or the target spec changes its anchors around, etc). So giving an escape hatch is definitely always necessary.  Might be willing to make it more obvious you're doing something "wrong", tho.

> It sounds like some of this is just working around #792 and #786 ?

Given that autolinking (and this aspect of it in particular) dates back to the start of Bikeshed, no, it's not working around those issues.  (Which are largely a result of a non-Bikeshed spec using <dfn>s in ways that Bikeshed/Shepherd don't quite agree with.)

> Idle thought: I wonder why https://streams.spec.whatwg.org/#index-defined-elsewhere doesn't use <dfn> but HTML's https://html.spec.whatwg.org/multipage/infrastructure.html#dependencies does. They seem semantically the same.

Because they're two different specs using markup in different ways? In particular, I think HTML uses those external definitions to enable a half-manual sort of external autolinking in Wattsi, right?  In Bikeshed, the index is just a companion to the internal definition index, and neither of those indexes contain definitions; one just points to the definition internally, the other externally.

In other words, I think Bikeshed is doing things correctly; HTML's use of DFN in the index is a Wattsi-ism that is correct locally but weird globally.

> public: designed to be called from other specs.
> descriptive: only designed to be "called" from the local spec, but can be used as part of another spec's explanatory text.
> private: only designed to be used from the local spec. For example, some locally-used terms might be too short to be used from elsewhere.

That's basically already captured, with a combination of `noexport` and `local-lt` - "public" is anything exported, "descriptive" is anything unexported, and "private" is the term you use in `local-lt`, which is only usable locally and not collected by Shepherd (so it never enters Bikeshed's databases). `local-lt` is intended for *precisely* the purpose of "I want to use a short term that would be bad to export globally, here's a longer version of the name for other people to use".

The only possibility not hit currently is something you really, truly only want to refer to locally; you can't currently use `<dfn local-lt>` without also happening to create a global term as well.  If it's important enough to be referred to locally, tho, it's probably actually in your "descriptive" category, and possibly appropriate for other specs to refer to.

Anyway, this is giving me some action items and things to think on:

* [ ] Provide some way to annotate a DFN to indicate that it's *not* a definition that Bikeshed should pay any attention to.  (Probably by just looking for an empty `lt`.)
* [ ] Consider how better to express the public/private distinction, and see if I can migrate things over to that automatically (presumably by making everything "public" by default, as the "dfns are unexported" thing is somewhat confusing and unintuitive).
    * [ ] Figure out some UI that, without being too annoying, bugs people if they're not marking things appropriately. Maybe a `bikeshed source` command that'll annotate everything as public, and the error message says "hey, looks like you've got X definitions not explicitly marked; mark them, or run `bikeshed source publicize-everything` to make everything explicitly public"?

**Comment by jyasskin** [comment]
@annevk Do you have some good examples of terms that should only be used from non-normative content? You mentioned ["navigate"](https://html.spec.whatwg.org/multipage/browsers.html#navigate) before; are there others?

**Comment by annevk** [comment]
E.g., in Fetch terms like "main fetch", "basic fetch", "HTTP-network fetch" are implementation details. They are private to Fetch. Standards still want to reference sometimes from a non-normative context to point out their implications on their algorithms (e.g., CSP does this).

Now, another thing. The problem with `<a spec=x>term</a>` to override `noexport` is that you also need the former when several standards use the same term. E.g., I use `<a spec=webidl>identifier</a>` here and there. So that means you cannot use the presence of the `spec` attribute coupled with it appearing in a normative context as an error.

I didn't mention that in OP, but that was part of my thinking when filing this, that the system is not completely thought out.

**Comment by annevk** [comment]
It would be really great if we could make some progress here. This is still a problem and we're starting to think about exposing public/private more clearly through stylesheets.

Another place where I think it would make sense to expose the difference is indexes. That way you can see at a glance which of your terms are public/private and whether or not they should be.

A thing @domenic came up is that perhaps we could make export/noexport public/private, but we'd additionally support a export+private combination for the couple times where folks want to reference private terms for some reason.

**Comment by tabatkins** [comment]
> but we'd additionally support a export+private combination for the couple times where folks want to reference private terms for some reason.

Can you elaborate on this more?

**Comment by tabatkins** [comment]
(I'm very open to abandoning export/noexport entirely; given that nearly everything is exported, and most unexported "dfn" terms are accidental, it's not serving its purpose right now.  Moving to a clearer public/private distinction and making it required would be nice. We just need to be really clear on what the desired model is before I force people into it.)

**Comment by annevk** [comment]
@tabatkins basically what I wrote in https://github.com/tabatkins/bikeshed/issues/797#issuecomment-243341658. "Main fetch" is private to Fetch, but is currently exported so another standard can explain in a note how its algorithms are affected by it (CSP and Mixed Content for instance have rather tightly coupled integration).

**Comment by tabatkins** [comment]
Ok, so that sounds like just a two-state public/private, where private can only normatively be referenced from inside the defining spec, but can informatively be referenced from anywhere. Right?

Are we *certain* that we want to actually make it impossible to autolink to a private thing normatively? I'm gonna be annoyed if it turns out a few weeks after I do this that y'all want some "friend" specs that can link each other's private state.

**Comment by annevk** [comment]
Well, we definitely want to give access to internal slots and such, but I guess those would simply be public, even though they're private state from the perspective of JavaScript.

We have some cases where something is public, but really only for one consumer, but those I tend to call out with a note of sorts as well. I think the distinction you make should be sufficient and will be an improvement over the status quo.

**Comment by domenic** [comment]
I'm not sure informatively referenced is correct, e.g. for monkeypatches or statements of fact.

I think what would make more sense to me is something similar to the current model in terms of how you reference these things: you need an explicit `spec=fetch`, or `yesireallymeantoaccesstheprivatething` attribute or some metadata block listing the private things you expect to access.

I'm on board with flipping things to public-by-default though.

Will this change span Bikeshed and Shepherd, or will we keep using Shepherd's data-export="" attributes?

**Comment by tabatkins** [comment]
Domenic: So it sounds like you want *precisely* the current model, just possibly with things defaulting to public?  Or maybe instead just start being stricter about marking everything as export or noexport?

**Comment by annevk** [comment]
I think with the current model if you're explicit about a private thing, it can still match a public thing. That should probably stop too. Maybe as simple as making public/private part of the key.

**Comment by tabatkins** [comment]
Oh, hm, so allow a public and private thing to overlap in name?

**Comment by annevk** [comment]
Or maybe that should just error to begin with. Dunno. I guess it was mostly HTML that had that problem and it's mostly been resolved.

**Comment by domenic** [comment]
> Domenic: So it sounds like you want precisely the current model, just possibly with things defaulting to public? Or maybe instead just start being stricter about marking everything as export or noexport?

Yeah, I guess so. Re-reading my earlier responses in the thread, that's pretty much what I said in https://github.com/tabatkins/bikeshed/issues/797#issuecomment-243180290, I now see.

**Comment by tabatkins** [comment]
Okay, so here's a draft plan:

1. Introduce a new `public`/`private` attribute pair for dfns. (The public API would be `data-dfn-public`/etc, but in Bikeshed it'll just be the word.)
2. Require public/private on *every* dfn. Fatal error if missing.
3. Public definitions can be autolinked normally. Private definitions can be autolinked normally from within non-normative sections. To link to a private definition from a normative section, require some obvious i-know-what-i'm-doing attribute to the autolink or a direct parent; this'll *only* handle this case, not pull double-duty like `spec` does currently.  (Probably just a `private-dfn` attribute on the autolink?)
4. Add some opt-out metadata to help with migration, that auto-applies public/private based on the current export/noexport rules. (But probably permanently logs a warning while in use.)
5. Just assume that everything is public otherwise, for specs that aren't using the new markup at all.

How does this sound?

**Comment by annevk** [comment]
That sounds pretty good. For IDL there's probably a fine line to walk, I think we actually want interface members to be "private" as other specifications shouldn't use them directly to define their features, but you're right that interfaces names have to be public (and globally unique).

**Comment by domenic** [comment]
That sounds pretty good. I'm curious though how this affects the dfn contract and Shepherd database, especially since HTML will be continuing to author all its attributes manually. Will we have to mass-update HTML from data-export="" to data-dfn-public="" or similar? Or does that fall under (6)?

I'm not sure about interface members. I might be conservative and default them to private for now, as all the legit uses I can think of will be in non-normative text, but I might be wrong and people want to refer to interface members all the time.

Looking through the other definition types: IDL arguments should definitely be private; IDL stringifiers and serializers should probably be private?; I'm not really sure what an iterator/maplike/setlike dfn is but they might also be private?

**Comment by tabatkins** [comment]
Yeah, stringifiers/serializers are just methods with magic names, so they should act the same as any other method.

That said, I'm not sure I agree that arguments/methods should be private.  People should definitely be linking to the slot/algorithm definitions instead, but unless/until WebIDL defines that there are automatically backing slots for all them, linking to the attr/method and referring to the private backing slot (like I do in [TypedOM](https://drafts.css-houdini.org/css-typed-om/#dom-cssnumericvalue-add) is probably the best I can do automatically.  Maybe we should prioritize that change?

(Basically, since you *almost 100%* want to refer to the backing slot / backing algorithm, not the actual property of the given name on the object, we can probably default to interpreting references to an attr/method to be talking about the backing slot/algo instead, and in the super-rare cases you actually want to access an object's property, you can do that with ES-ish syntax.)

I'm... actually not sure what iterator/maplike/setlike are supposed to do.  I might have intended them to link to the map entries/etc definition for the interface?  Given that widlparser doesn't ever output them, I think I can just drop those from Bikeshed.

**Comment by annevk** [comment]
> we can probably default to interpreting references to an attr/method to be talking about the backing slot/algo instead

In a way that seems reasonable, but I'd rather we not, as it's another level of indirection readers will have to become familiar with. I'd rather we describe things as they actually are.

**Comment by domenic** [comment]
I think arguments should definitely be private though, regardless of what we do about attributes/operations.

**Comment by tabatkins** [comment]
Yeah, I agree about arguments.

@annevk Hmm. I guess it's fine for me to call attrs/methods private for now. Outside of a given spec you should rarely be calling into another object's slots/algos anyway.  I'll just jump up the WebIDL effort to make all attributes/methods *automatically* backed by a slot/algo, with Bikeshed supporting that automatically. (Then the TypedOM stuff would just switch to referring to `{{[[unit]]}}` etc, instead of `{{unit}} internal slot` like I'm doing right now to handwave it.)
